### ISSUE TITLE
Immediately destructuring a spread union allows access to properties not present in all union constituents

### ISSUE THREAD

**Issue by jcalz** [microsoft/TypeScript#43174]
# Bug Report

### 🔎 Search Terms

spread, destructuring, union

### 🕗 Version & Regression Information

- This changed between versions 3.5.1 and 3.6.2.  Specifically it changed with #31711.  It is still present in 4.2.3 and in nightly.

### ⏯ Playground Link

[Playground link](https://www.typescriptlang.org/play?ts=4.2.3#code/JYOwLgpgTgZghgYwgAgILIN7LgLmQZzClAHNkBfAWAChRJZEUAhTZAIzxAFcBbN6CjRowuIBGGAB7EMhiTJACgAeedAB9kTAJSYayfcgTTCrOBWQBeVgDpbS8wHoHyEJOTQokqAH4AhMj0DJ0NjMGw8QmIQMg1RABMIGFAIOMD9OG9rMEkAVQAHPOgAYTh8CAUtAG4aKmoaIxATeyssXGQARgAadjwAJkFqOUUlKuRguCzcguLSlGB8F0kwsxExCWlu+Tz8IUHRcSkZId7lVWQNbV1qAxDGsNbzKyVKsecPL38ABU9CqDAAT2QAHI4EDkHFJBAFq4whAlPMwtJkADCsD1JogdYartVgckUMAMynNDnTQ6DBpW4mABulhsdgGNwaJge5Dp1JewXeUC+P2gAOBoPBkOhS3c8JMSJRKCBrQiRFILzZGiwHBcvH4UCVQJqQA)

### 💻 Code

```ts
interface A { a: string }
interface B { b: number }

function foo(x: A | B) {
    const { a } = { ...x } // no error?! 
    // const a: string | undefined
    a?.toUpperCase();
}

const x = { a: 1, b: 2 }
foo(x); // a.toUpperCase is not a function, oops
```

### 🙁 Actual behavior

There is no error; `a` is inferred to be of type `string | undefined`.

### 🙂 Expected behavior

I expected an error on destructuring assignment to `a` saying `Property 'a' does not exist on type 'A | B'`, as you would get with the seemingly equivalent construction without spreading

```ts
const { a } = x; // error! Property 'a' does not exist on type 'A | B'.
```

or the also-seemingly-equivalent construction with an intermediate variable
```ts
const v = { ...x }
const { a } = v; // error! Property 'a' does not exist on type '{ a: string; } | { b: number; }'
```

---

I figure this is a side effect of #31711.  But is it considered a bug? a design limitation? or working as intended?  Comes from [this SO question](https://stackoverflow.com/questions/66557743/property-aaa-does-not-exist-on-type-bbb-error-from-tsc-typescript).
